src/Compilers/CSharp/Portable/SourceGeneration/SyntaxValueProviderExtensions.cs

```diff
@@ -0,0 +1,262 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Transactions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.SourceGeneration;
+
+using Aliases = ArrayBuilder<(string aliasName, string symbolName)>;
+
+internal static class SyntaxValueProviderExtensions
+{
+    /// <summary>
+    /// Returns all syntax nodes of type <typeparamref name="T"/> if that node has an attribute on it that could
+    /// possibly bind to the provided <paramref name="attributeName"/>. <paramref name="attributeName"/> should be the
+    /// simple, non-qualified, name of the attribute, including the <c>Attribute</c> suffix, and not containing any
+    /// generics, containing types, or namespaces.  For example <c>CLSCompliantAttribute</c> for <see
+    /// cref="System.CLSCompliantAttribute"/>.
+    /// <para/> This provider understands <see langword="using"/> aliases and will find matches even when the attribute
+    /// references an alias name.  For example, given:
+    /// <code>
+    /// using XAttribute = CLSCompilation;
+    /// [X]
+    /// class C { }
+    /// </code>
+    /// Then
+    /// <c>context.SyntaxProvider.CreateSyntaxProviderForAttribute&lt;ClassDeclarationSyntax&gt;("CLSCompliant")</c>
+    /// will find the <c>C</c> class.
+    /// </summary>
+    public static IncrementalValuesProvider<T> CreateSyntaxProviderForAttribute<T>(this SyntaxValueProvider provider, string attributeName)
+        where T : SyntaxNode
+    {
+        // Create a provider that provides (and updates) the global aliases for any particular file when it is edited.
+        var individualFileGlobalAliasesProvider = provider.CreateSyntaxProvider(
+            (n, _) => n is CompilationUnitSyntax,
+            (context, _) => GetGlobalAliasesInCompilationUnit((CompilationUnitSyntax)context.Node));
+
+        // Create an aggregated view of all global aliases across all files.  This should only update when an individual
+        // file changes its global aliases.
+        var allUpGlobalAliasesProvider = individualFileGlobalAliasesProvider
+            .Collect().Select((arrays, _) => GlobalAliases.Create(arrays.SelectMany(a => a.AliasAndSymbolNames).ToImmutableArray()));
```

Comment:
note: it's unclear to me how Collect works.  My hope is that it doesn't rerun later parts if it sees that all its inputs have retained value-equality.

Reply:
`Collect` is implemented using the `BatchNode<T>` which does have caching semantics. At the same time we should be able to verify this with testing correct?

Reply:
yeah, i believe this is workign properly (and all the tests validate each stage of hte pipeline.  wouldn't mind additional confirmation though.  I was askign this prior to getting all the tests and learning a lot that way.